src/frequenz/sdk/microgrid/_graph.py

```diff
@@ -352,6 +495,179 @@ def validate(self) -> None:
+        return (
+            component.category == ComponentCategory.METER
+            and len(successors) > 0
+            and all(self.is_ev_charger(successor) for successor in successors)
+        )
+
+    def is_ev_charger_chain(self, component: Component) -> bool:
+        """Check if the specified component is part of an EV charger chain.
+
+        Args:
+            component: component to check.
+
+        Returns:
+            Whether the specified component is part of an EV charger chain.
+        """
+        return self.is_ev_charger(component) or self.is_ev_charger_meter(component)
+
+    def is_battery_inverter(self, component: Component) -> bool:
+        """Check if the specified component is a battery inverter.
+
+        Args:
+            component: component to check.
+
+        Returns:
+            Whether the specified component is a battery inverter.
+        """
+        return (
+            component.category == ComponentCategory.INVERTER
+            and component.type == InverterType.BATTERY
+        )
+
+    def is_battery_meter(self, component: Component) -> bool:
+        """Check if the specified component is a battery meter.
+
+        This is done by checking if the component has only battery inverters as
+        its successors.
+
+        Args:
+            component: component to check.
+
+        Returns:
+            Whether the specified component is a battery meter.
+        """
+        successors = self.successors(component.component_id)
+        return (
+            component.category == ComponentCategory.METER
+            and len(successors) > 0
+            and all(self.is_battery_inverter(successor) for successor in successors)
+        )
+
+    def is_battery_chain(self, component: Component) -> bool:
+        """Check if the specified component is part of a battery chain.
+
+        Args:
+            component: component to check.
+
+        Returns:
+            Whether the specified component is part of a battery chain.
+        """
+        return self.is_battery_inverter(component) or self.is_battery_meter(component)
+
+    def is_chp(self, component: Component) -> bool:
+        """Check if the specified component is a CHP.
+
+        Args:
+            component: component to check.
+
+        Returns:
+            Whether the specified component is a CHP.
+        """
+        return component.category == ComponentCategory.CHP
+
+    def is_chp_meter(self, component: Component) -> bool:
+        """Check if the specified component is a CHP meter.
+
+        This is done by checking if the component has only CHPs as its
+        successors.
+
+        Args:
+            component: component to check.
+
+        Returns:
+            Whether the specified component is a CHP meter.
+        """
+        successors = self.successors(component.component_id)
+        return (
+            component.category == ComponentCategory.METER
+            and len(successors) > 0
```

Comment:
Can this be `==1`?

Reply:
yes,  but the component graph can be generic, doesn't have to care about the tighter requirements of the formula generator (which might go away, especially if we encounter some complicated setups)